docs/PROVIDER.md

```diff
@@ -0,0 +1,110 @@
+
+The clusterstackrelease-controller will search for the `ProviderClusterStackRelease` object that is referenced in the spec of `ClusterStackRelease` and check whether `status.ready` is `true`. If it is not, it will wait. Only if the provider-specific jobs are completed and the `status.ready` boolean is set on `true`, the reconcile loop will proceed and apply the required Cluster API objects in the management cluster.
+
+It makes sense to wait, because otherwise the `ClusterClass` object would be present that allows the user to use the respective release of the cluster stack. However, if the node images are not ready yet, the cluster cannot start successfully. Therefore, to create the `ClusterClass` object is the last step to be done.
+
+## Provider Contract
+
+A Cluster Stack Provider Integration ensures that the node images that are released e.g. as build information in a release of a cluster stack, are accessible and ready to use. A user who creates a workload cluster via the the `ClusterClass` object of a release of a cluster stack should have all node images provided.
+
+
+A Cluster Stack Provider should be built as a separate Kubernetes Operator.
+
+### Custom resources
+
+#### ProviderClusterStackRelease resource
+
+The `ProviderClusterStackRelease` resource must have a `status` field `ready` (boolean), indicating that the provider-specific tasks (notably providing node images) are completed for the respective release.
+
+#### ProviderClusterStackReleaseTemplate resource
+
+For a `ProviderClusterStackRelease` resource, there must be a corresponding template resource following this pattern:
+
+```
+// ProviderClusterStackReleaseTemplateSpec defines the desired state of ProviderClusterStackReleaseTemplate.
+type ProviderClusterStackReleaseTemplateSpec struct {
+	Template ProviderClusterStackReleaseTemplateResource `json:"template"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=providerclusterstackreleasetemplates,scope=Namespaced,categories=cluster-stack,shortName=pcrt
+// +kubebuilder:storageversion
+
+// ProviderClusterStackReleaseTemplate is the Schema for the providerclusterstackreleasetemplates API.
+type ProviderClusterStackReleaseTemplate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ProviderClusterStackReleaseTemplateSpec `json:"spec,omitempty"`
+}
+
+type ProviderClusterStackReleaseTemplateResource struct {
+	// +optional
+	ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
+	Spec ProviderClusterStackReleaseSpec `json:"spec"`
```

Comment:
I think you can have a look at CAPI's docs for provider implementations. They have the same structure. I cannot add the information you requested, because there is literally no pre-defined structure and we don't need to know anything about `ProviderClusterStackReleaseSpec`. 

I conclude for myself that this part is misleading more than helping, so we can think of removing this code example alltogether.

Reply:
@janiskemper I agree that these code examples do not help the reader so much.

Reply:
I do not agree about the non-common structure issue. Please have a look at e.g. `runtime.RawExtension` which is exactly for that use case. You can give a specific k8s type with provider specific sections in place where providers can add their own structure needed.

[1] https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension
[2] https://github.com/kubernetes/apimachinery/issues/102

Reply:
so can you give me an example from the Cluster API context where this is used?

Reply:
See [1] this discussion for a good overview of the Cluster API decision. In short they replaced the need of provider specific fields by requiring the well known `<Provider><Type>` mechanism and referencing it.

Going one step back the first question to find out if `ProviderClusterStackRelease` is really expected to be named that way or if `<Provider>ClusterStackRelease` like `OpenStackClusterStackRelease` is meant instead. If the later is the case than no `RawExtension` use case exists but then it's the providers duty to report if a release is ready or not by patching the respective `status` field. That would be an excellent example of an common `cluster-stack-provider` Go package use case.

[1] https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20190709-cluster-spec-crds.md

Reply:
Maybe I completely misunderstand you, but I would say we name it exactly in this way, e.g. that for OpenStack it will be OpenStackClusterStackRelease.

Reply:
@matofeder after the conversation I had with multiple other people involved in the project of implementing a provider, I realized that this code example is actually really necessary. I have added a sentence now that `ProviderClusterStackReleaseSpec` has to be defined by the person implementing it to clarify it a bit better.

Reply:
@janiskemper thx

Reply:
Thanks for the additional sentence, @janiskemper, should make it more clear. I was wondering whether it makes sense to even write out the example, in this case
```
The prefix `Provider` should be replaced with the respective provider to build, e.g. for `Docker` it would be `DockerClusterStackReleaseSpec`.
```
These are details now, but would further limit the scope for interpretation.

Is there anything to add from you @NotTheEvilOne?
Otherwise we should mark this conversation as resolved

Reply:
The changes added here are sufficient, yes.

Reply:
and yes, @jschoone I can incorporate your changes!